doc: the Dune Mental Model

emillon opened the pull request.

emillon committed:
doc: the Dune Mental Model

Signed-off-by: Etienne Millon <me@emillon.org>
Co-authored-by: Andrey Mokhov <andrey.mokhov@gmail.com>
Co-authored-by: Stephen Sherratt <stephen@sherra.tt>
Co-authored-by: Marek Kubica <marek@xivilization.net>

emillon commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+of *t*. (Dune will raise an error in other cases: if *t* is both in the source
+tree and the target of a rule, or if it is neither)
+
+In the case of our example, if we call ``dune runtest``, Dune will consider all
```
this paragraph is a mouthful, but a bit on purpose. I wonder if it could work as a nested bullet list, but this feels too technical

snowleopard replied:
I've found this paragraph too mechanistic and not very insightful. Maybe we can even skip it? Or, hide it and include the bullet list into a hidden section if anyone would like to step though it.

gridbugs replied:
I personally find explicitly stepped examples like this very helpful when learning how to use new tools.

snowleopard replied:
OK, let's keep it. Looks like most people want it.

snowleopard commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+
+Note that in this example, some files are targets of a rule and dependencies of
+another (``.cmx`` files). We are unlikely to ever interact with them directly,
+so it can also be useful to think of the ``(executable)`` stanza as acting as a
```
I'm curious why you find it useful to think of this as just one rule. From the point of view of Dune's performance/cache model, it's pretty important that these rules are separate. (On the other hand, bundling these separate rules into one is sometimes appropriate and internally we have some optimisations that are actually working this way.)

Leonidas-from-XIV replied:
As a second opinion, I find it useful to think about it that way, since as a Dune user one rarely if ever need to think about `.cm*` files, so the intermediate rules aren't all that interesting.

I would even consider making this thinking the starting point since that's sort of the intuition and then explain the workings in more detail later.

emillon replied:
This comes from this graph from the presentation (omitting the graphs in a first version is a deliberate design decision, so that the text has to be clear and stays accessible)

```mermaid
graph LR
  tool.exe;
  config.ml;
  main.ml;
  subgraph "executable: tool"
    bm[/compile Main/] --> main.cmx --> l[/link executable/]
    bc[/compile Config/] --> config.cmx
  end
  
  l[/link executable/] --> tool.exe
  config.cmx --> l
  config.ml --> bc
  main.ml --> bm
```

Here what I meant is that the collection of rules from the executable stanza act as an opaque box with the `ml` files as dependencies and `tool.exe` as a target. I think that's useful to think it that way but I shouldn't call that a rule.

snowleopard replied:
I like the graph! My feeling is that it's sufficient to say that the highlighted group of rules corresponds to the `executable` stanza and leave it at that (i.e. no need to talk about thinking about them as one rule).

snowleopard commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+-----------------
+
+Some rules do not produce any output file, but we're still interested in
+running an action. A test is a good example: we want the build process to exit
```
```suggestion
running their actions. A test is a good example: we want the build process to exit
```

snowleopard commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+
+- reads its dependencies: ``tool.exe`` and ``testdata.txt``
+- writes no targets
+- by running its action: ``(run tool.exe testdata.txt)``
```
"by running" reads odd here, maybe "and runs"?

emillon replied:
that's intended to be a callback to the blockquote at the beginning, "a rule reads dependencies and writes targets using an action". I tried to rephrase it.

snowleopard commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+- reads its dependencies: ``tool.exe`` and ``testdata.txt``
+- writes no targets
+- by running its action: ``(run tool.exe testdata.txt)``
+- (but it is attached to ``runtest``)
```
```suggestion
- (and it is attached to ``runtest``)
```

snowleopard commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+running an action. A test is a good example: we want the build process to exit
+with an error code if the action fails.
+In that case, the rule does not have targets, but we "attach" it to an
+:term:`alias`, ``runtest`` in this case.
```
Maybe it's worth saying why we attach such actions to aliases: if we didn't, there would be no way to force executing them, since they have no targets. We could also remind the readers here that Dune is lazy: it only executes actions that are required to build the requested build goals.

gridbugs replied:
Agreed that it would be helpful to explain why the alias is necessary. My question reading this part is: Since dune doesn't run actions from rules whose targets all exist, and this rule has no targets, why does it ever run actions of rules with no targets? Is this a special case for rules with no targets?

Leonidas-from-XIV replied:
Doesn't it actually have some sort of hidden target? After all, if the test successfully finishes, it doesn't get re-run again.

I tripped up on this with e.g. property-based tests which don't have a reproducible outcome, once it succeeds you need `--force` to re-run it as there's no change in dependencies.

emillon replied:
I'm not sure about the details, but I think that the implementation of aliases is based on "tokens" (stamp files) whose name is based on the digest of their dependencies

snowleopard replied:
We shouldn't include such implementation details in the mental model. Stamp files are likely to disappear at some point (maybe even soon).

snowleopard replied:
(I'm thinking of nuking stamp files became writing a file on disk is a pretty inefficient way to cache an action execution.)

snowleopard commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+Now that we have seen how Dune sets up rules for a project, let's see how it
+executes a ``dune build`` command.
+
+Dune can build targets: either files, or aliases. ``dune build tool.exe`` will
```
```suggestion
Dune can build *goals*: either files or aliases. ``dune build tool.exe`` will
```
I think "goal" is the usual term to denote things that, e.g. can appear on the command line.

gridbugs replied:
I've never encountered the word "goal" in dune or other build systems. To avoid introducing a new piece of terminology, what about rewording it to:
:
```suggestion
Dune can build files and aliases. ``dune build tool.exe`` will
```

Leonidas-from-XIV replied:
Likewise I don't think I ever saw them referred to as goals. I was thinking of what Make calls them and they are usually "Makefile targets". Which are after all very close to files that we build (targets) and aliases (what Make calls "phony targets" and hides behind terrible syntax).

(On the other hand "Dune can build [...] aliases" is oddly worded, maybe "build files" and "evaluate aliases"?)

snowleopard replied:
I like @gridbugs's version. We use "goals" in implementation but happy to leave it there.

I also think that saying "build files and aliases" is fine because it directly corresponds to the CLI `dune build file @alias`.

snowleopard commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+alias (``dune runtest`` is a shortcut for ``dune build @runtest``).
+
+To build the target *t*, Dune will first check is *t* is in the source tree. In
+that case, it will copy it to the ``_build`` directory. Otherwise, it will
```
I think "it will copy it to the `_build` directory" may be leaking some unnecessary implementation details. Without thinking deeply about how Dune builds are working, copying sources into the `_build` directory might seem unnecessary/confusing. In fact, internally, this will soon not be true in some cases.

snowleopard replied:
So, maybe, "In this case, there is nothing to do" instead?

snowleopard commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+To build the target *t*, Dune will first check is *t* is in the source tree. In
+that case, it will copy it to the ``_build`` directory. Otherwise, it will
+check if it is the target of a rule *r*. In that case, it will build all
+dependencies of *r* and if any dependency has changed, it will run the action
```
> and if any dependency has changed

This is a bit vague (changed since when?). I find it easier to separate caching into a completely separate concern, which only applies to executing actions. Like this:

When Dune is about to execute an action, it checks (in various caches) if it executed it before on the same set of dependencies, and, if yes, it can skip executing it and reuse the previous result.

snowleopard commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+In the case of our example, if we call ``dune runtest``, Dune will consider all
+rules attached to the ``runtest`` alias. This is just the integration test
+rule. It needs to build its dependencies, ``tool.exe`` and ``testdata.txt``.
+The latter is present in the source tree, so it just needs to be copied.
```
> so it just needs to be copied

Again, I think this detail can be skipped.

gridbugs commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+The Dune Mental Model
```
I like "The Dune Mental Model".

gridbugs commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+How Dune Works
+--------------
+
+The building brick of Dune is the *rule*:
```
I think "building block" is the more commonly used term.

emillon replied:
too much lego at home

gridbugs commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+
+- there's a CLI tool written in OCaml.
+- it has some build-time configuration stored in ``config.json``.
+- it has an integration test which consists in running in it with
```
```suggestion
- it has an integration test which consists of running in it with
```

emillon replied:
I believe that "consist of" defined the components of something , while "consist in" introduces a definition?
Maybe
"it has an integration test, in which the tool is executed with `testdata.txt` as input"?

gridbugs commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+   (name tool)
+   (modules main config))
+
+Here, we use Dune's abstractions. It knows about the OCaml compilation model:
```
To me this reads more clearly if the first "it" is replaced with "dune".
```suggestion
Here, we use Dune's abstractions. Dune knows about the OCaml compilation model:
```

gridbugs commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+build ``tool.exe``, while ``dune build @runtest`` will build the ``runtest``
+alias (``dune runtest`` is a shortcut for ``dune build @runtest``).
+
+To build the target *t*, Dune will first check is *t* is in the source tree. In
```
I think it would help to clarify that this paragraph is only about files and not aliases. It might also be helpful to explain under what conditions does dune actually build an alias (I don't actually know the answer to this). What if rule has both an alias and target files and all the target files are present?

gridbugs commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+Dune's underlying model is based on rules. Stanzas are high-level constructs
+that can generate multiple rules, that are not always visible.
+
+To build a target, Dune looks for the rule that produces that target and makes
```
```suggestion
To build a target or alias, Dune looks for the rule that produces that target and makes
```
Not sure this change is necessary, but since the `(target ...)` syntax is used for file targets and `(alias ...)` is used for aliases, using the word "target" here might suggest that the behaviour when building an "alias" is different.

gridbugs replied:
Oh actually this paragraph reads as though it's only about file targets, in which case I'd say leave it as such and add an additional sentence summarizing how dune builds aliases.

gridbugs commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+
+To build a target, Dune looks for the rule that produces that target and makes
+its way back to source files. This explains why it is generally fine with
+"dynamic" dependencies, but needs "static" targets.
```
I think the detail about dynamic dependencies and static targets should either be skipped here or elaborated upon further outside the conclusion. This page doesn't talk about dynamically generating the dependencies of a rule so I think it's a bit out of place here, and a fairly high-level concept to introduce at this point in the page.

gridbugs commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+its way back to source files. This explains why it is generally fine with
+"dynamic" dependencies, but needs "static" targets.
+
+Rules define a graph which models precise relations between files. Most of the
```
```suggestion
Rules define a directed acyclic graph which models precise dependencies between files. Most of the
```

emillon replied:
I like the DAG part, but if we say "dependencies between files", this reuses the "dependency" word for something else. Maybe "which models dependency relations"?

gridbugs commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+-------------------------
+
+Now that we have seen how Dune sets up rules for a project, let's see how it
+executes a ``dune build`` command.
```
I think it would also be helpful to explain what happens when running `dune build` with no arguments. What targets/aliases does it attempt to build in that case?

gridbugs commented:
Thanks for making this @emillon! I think this stands to have a big impact on helping people understand dune.

Leonidas-from-XIV commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+
+It is not strictly necessary to understand Dune's underlying model to use it;
+but knowing how it works under the hood will help understand some errors and
+what's possible with Dune.
```
I think it also makes writing rules easier if you know what you're doing instead of (I assume many like me started out just copying boilerplate from somewhere else).

Leonidas-from-XIV commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+
+    A *rule* reads *dependencies* and writes *targets* by executing an *action*.
+
+When ``dune build`` is executed, it will first read ``dune`` files to determine
```
```suggestion
When ``dune build`` is executed, it will first read the project's ``dune`` files to determine
```

As it feels a bit like there's a word missing.

Leonidas-from-XIV commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+  - and write its output: ``config.cmx``
+  - using an action: ``(run ocamlopt -c config.ml)``
+
+- one rule to compile the ``tool.exe`` executable:
```
In theory "compile" is always correct in the broader sense of the word but I agree that as a reader I would expect "link" here, even if this is not necessarily what `ocamlopt` is doing.

But maybe this opinion is influenced too much by the C point of view.

emillon replied:
yes, I meant link here

Leonidas-from-XIV commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+tree and the target of a rule, or if it is neither)
+
+In the case of our example, if we call ``dune runtest``, Dune will consider all
+rules attached to the ``runtest`` alias. This is just the integration test
```
```suggestion
rules attached to the ``runtest`` alias. In this case it is just the integration test
```

Leonidas-from-XIV commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,175 @@
+In a nutshell: we recursively copied all the dependencies of the test rule, and
+executed the rules in the correct order.
+
+This is a "cold build", where nothing was previously built. But note that if we
```
```suggestion
This is a "cold build", where there were no previous build artifacts. Note that if we
```

Leonidas-from-XIV reviewed (commented):
I think this would easily be the page of the documentation I would suggest people to read the most.

Added some of my comments and :+1: to the previous comments that this document has gotten.

ejgallego commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,174 @@
+The building block of Dune is the *rule*:
+
+    A *rule* reads *dependencies* and writes *targets* by executing an *action*.
+
```
Sometimes when I have been explaining the mental model to people I've used:

```
{ deps } ---[action]---> { targets }
```

emillon commented:
Folks, thanks a lot for your feedback. I've reworked the document a bit, in particular:

- I added a "what to build" section. the goal is to explain that we have (A) a set of rules derived from dune files (B) a "goal" (I've not used the term in the doc) and that (C) using A and B as inputs, an algorithm will do the right thing
- in part "C" (the build algorithm) I've detailed what it means to build a file, build an alias, and execute a rule.

Let me know if you have additional remarks.

Leonidas-from-XIV commented on doc/explanation/mental-model.rst:
```diff
@@ -0,0 +1,197 @@
+with an error code if the action fails.
+In that case, the rule does not have targets, but we "attach" it to an
+:term:`alias`, ``runtest`` in this case. This gives us a way of requesting this
+rule to be executed. As we are about to see, rules are executed lazily by
```
While this sort of implies it is possible I think it would be nice to state directly that more than one rule can be attached to an alias and in fact `runtest` has lots of rules attached to them.

(That's one way my understanding of aliases as "phony Makefile targets" did not match up)

emillon replied:
a "dune for make users" guide could be useful indeed!

Leonidas-from-XIV reviewed (approved):
I think it looks good and further improvements can be added when necessary in the future anyway.
